### PR TITLE
QUIC Send Stream Management

### PR DESCRIPTION
**Pull request by hlandau**
For usage information and discussion, see `quic_stream.h`.

The PN set structure which was included as part of the ACKM has been refactored out into a generic 64-bit UINT_SET, as it was also needed by this code.

### CODE REVIEW

File: ssl/quic/quic_stream.c
Comment (paulidale) [inline ssl/quic/quic_stream.c:302]: hmmmmm.  We don't put asserts in the library.

File: ssl/quic/quic_stream.c
Comment (paulidale) [inline ssl/quic/quic_stream.c:328]: blank line style nit.

File: ssl/quic/quic_stream.c
Comment (hlandau) [reply ssl/quic/quic_stream.c:302]: Wait, what? I've been using these extensively. libssl looks full of them.

File: ssl/quic/quic_stream.c
Comment (hlandau) [reply ssl/quic/quic_stream.c:75]: The assumption on my part is that the buffer we are copying from has been provided by the user in a call to `SSL_write`. So it's one copy to internal staging prior to encryption, and then a copy-and-encrypt.

So this is two-copy. We obviously can't support single copy without having the user use an API other than SSL_write (e.g. an API where we provide a buffer for them to fill, or where they lend us a buffer and we tell them when we're done with it via a callback, etc.).

So this is the best we can do. I'll add a note.

File: include/internal/quic_stream.h
Comment (hlandau) [reply include/internal/quic_stream.h:49]: Got any suggestions?

File: ssl/quic/quic_stream.c
Comment (mattcaswell) [reply ssl/quic/quic_stream.c:302]: > hmmmmm. We don't put asserts in the library.

Err...yes we do? Generally `ossl_assert` is preferred so that we do something sensible in a non-debug build. `OPENSSL_assert` should not be used in the library since that causes a crash even in a production build, which we should not do.

File: ssl/quic/quic_stream.c
Comment (hlandau) [reply ssl/quic/quic_stream.c:462]: If this happens the outcome is that it materialises as backpressure to the sender, which seems like a good way to handle this to me.

**Comment by hlandau** [comment]
Minor update.

Holding off on refactoring UINT_SET to use internal/list.h as I think I've identified some usabililty issues with it that need addressing.

**Comment by paulidale** [comment]
Issues with list.h or the UINT_SET?
I.e. do I have something to fix or not :smile:

File: ssl/quic/quic_stream.c
Comment (t8m) [reply ssl/quic/quic_stream.c:304]: We can use assert(). Why would not we? Of course, it should be used only at places where there is no way to return an error. Or to check invariants in debug builds that cannot be checked in production builds. Otherwise ossl_assert() should be used. We should never use OPENSSL_assert except for cases where we really want to check an invariant even in production build and where even returning an error could do more harm than terminating the process immediately - i.e., severe memory corruption detected.

**Comment by hlandau** [comment]
list.h. I'll probably raise a PR finessing it a bit at some point.

File: include/internal/quic_stream.h
Comment (t8m) [reply include/internal/quic_stream.h:49]: Yes, you do not need the ossl_ prefix for structures and other non-exported internal identifiers.

**Comment by hlandau** [comment]
Updated, rename.

File: ssl/quic/quic_stream.c
Comment (t8m) [inline ssl/quic/quic_stream.c:158]: Missing deallocation of `rnew.start`

File: ssl/quic/uint_set.c
Comment (t8m) [inline ssl/quic/uint_set.c:1]: Missing copyright header

**Review by t8m — CHANGES_REQUESTED** [review]
Just a few nits to fix and I'll approve. Great work!

**Comment by hlandau** [comment]
Updated.

**Review by t8m — APPROVED** [review]

**Comment by hlandau** [comment]
Rebased

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Ping again for the second review

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: include/internal/quic_stream.h
Comment (hlandau) [reply include/internal/quic_stream.h:104]: Maximum Datagram Payload Length, clarified

**Comment by hlandau** [comment]
Updated.

**Review by t8m — APPROVED** [review]

**Review by mattcaswell — APPROVED** [review]

**Comment by t8m** [comment]
Merged to master branch. Thank you.

**Comment by wan7teen-ping** [comment]
Hi I will help you with everything you need but you have to subscribe to
the youtube channel and I will fulfill all your requests

https://youtu.be/aIaqr4eq18g

Pada Rab, 7 Sep 2022, 9:40 PM Hugo Landau ***@***.***>
menulis:

**Pull request closed without merging**
